demos/demo_runner/iot_demo_afr.c

```diff
@@ -401,3 +405,201 @@ void vApplicationStackOverflowHook( TaskHandle_t xTask,
+        {
+            memcpy( chunk, pData, dataLength );
+            pCurrent = chunk;
+            pData += _MD5_CHUNK_LENGTH_BYTES;
+            dataLength = 0;
+        }
+        else
+        {
+            pCurrent = ( uint32_t * ) pData;
+            pData += _MD5_CHUNK_LENGTH_BYTES;
+            dataLength -= _MD5_CHUNK_LENGTH_BYTES;
+        }
+
+        for( i = 0; i < 64; i++ )
+        {
+            if( i < 16 )
+            {
+                F = ( B & C ) | ( ( ~B ) & D );
+                G = i;
+            }
+            else if( i < 32 )
+            {
+                F = ( D & B ) | ( ( ~D ) & C );
+                G = ( ( 5 * i ) + 1 ) % 16;
+            }
+            else if( i < 48 )
+            {
+                F = ( B ^ C ) ^ D;
+                G = ( ( 3 * i ) + 5 ) % 16;
+            }
+            else
+            {
+                F = C ^ ( B | ( ~D ) );
+                G = ( 7 * i ) % 16;
+            }
+
+            F = F + A + K[ i ] + pCurrent[ G ];
+            A = D;
+            D = C;
+            C = B;
+            B = B + LEFT_ROTATE( F, S[ i ] );
+        }
+
+        pOutput[ 0 ] += A;
+        pOutput[ 1 ] += B;
+        pOutput[ 2 ] += C;
+        pOutput[ 3 ] += D;
+    }
+}
+
+/*
+ * @brief Length of device identifier.
+ * Device identifier is represented as hex string of MD5 hash of the device certificate.
+ */
+#define AWS_IOT_DEVICE_IDENTIFIER_LENGTH ( _MD5_HASH_LENGTH_BYTES * 2 )
+
+/**
+ * @brief Device metrics name format
+ */
+#define AWS_IOT_METRICS_NAME "?SDK=" IOT_SDK_NAME "&Version=4.0.0&Platform=" IOT_PLATFORM_NAME "&AFRDevID=%.*s"
+
+ /**
+  * @brief Length of #AWS_IOT_METRICS_NAME.
+  */
+#define AWS_IOT_METRICS_NAME_LENGTH    ( ( uint16_t ) ( sizeof( AWS_IOT_METRICS_NAME ) + AWS_IOT_DEVICE_IDENTIFIER_LENGTH ) )
+
+/**
+ * @brief Unique identifier for the device.
+ */
+static char deviceIdentifier[ AWS_IOT_DEVICE_IDENTIFIER_LENGTH + 1] = { 0 };
+
+/*
+  @brief Device metrics sent to the cloud.
+ */
+static char deviceMetrics[AWS_IOT_METRICS_NAME_LENGTH + 1] = { 0 };
+
+
+static void _generateDeviceMetrics(void)
+{
+    size_t length;
+    length = snprintf( deviceMetrics,
+        AWS_IOT_METRICS_NAME_LENGTH,
+        AWS_IOT_METRICS_NAME,
+        AWS_IOT_DEVICE_IDENTIFIER_LENGTH,
+        deviceIdentifier );
+    configASSERT(length > 0);
+}
+
+/*
+ * @brief Generates a unique identifier for the device and metrics sent by the device to cloud.
+ * Function should only be called once at intialization.
+ */
+void generateDeviceIdentifierAndMetrics( void )
+{
+    const char *pCert = keyCLIENT_CERTIFICATE_PEM;
+    size_t certLength = strlen( keyCLIENT_CERTIFICATE_PEM );
+    uint8_t hash[ _MD5_HASH_LENGTH_BYTES ] = { 0 };
+    char *pBuffer = deviceIdentifier;
+    int i;
+    _generateHash( pCert, certLength, hash, _MD5_HASH_LENGTH_BYTES );
+
+    for( i = 0; i < _MD5_HASH_LENGTH_BYTES; i++ )
+    {
+        sprintf( pBuffer, "%02X", hash[ i ] );
```

Comment:
maybe snprintf? just because it is more secure....

Reply:
Fixed